### PR TITLE
refactor(protocol-engine): Implement `moveToCoordinates`

### LINKED ISSUE #10757: PE/JSONv6: Implement `moveToCoordinates` command
Implement the `moveToCoordinates` Protocol Engine command as it's already been specified in the JSON Protocol schema v6.

https://github.com/Opentrons/opentrons/blob/b8c7c8878a8c99ad2f2c5d6bc150faecb0b7f7d4/shared-data/protocol/schemas/6.json#L403-L431

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

This PR implements the Protocol Engine `moveToCoordinates` command. In the short term, this is necessary to fully implement the JSON Protocol v6 spec. In the longer term, it will be important in the implementation of Python Protocol API v3.

Closes #10757. See that ticket for the API specification.

# Changelog

* Implement the `moveToCoordinates` Protocol Engine command as previously specified.
* Fix a `float`->`int` truncation bug in the `JSONCommandTranslator`.
* Refactor the tests for `MotionView.get_waypoints_to_well()`.

# Manual testing

I think we currently have to test this by running commands on a real robot. And we have to do it over HTTP, since PAPIv3 isn't implemented enough to emit this command, and I assume Protocol Designer doesn't currently emit this command.

I've hacked together some scripts to help with this in Opentrons/otietalk#1.

Some interesting things to look out for:

* [x] When the pipette has a tip, the bottom of the *tip* should move to the given coordinates. When it doesn't, the bottom of the *nozzle* should move to the given coordinates.
* [x] When an 8-Channel pipette is used, the *back-most* channel should move to the given coordinates.
* [x] By default, the pipette should automatically move high enough to clear all loaded labware and modules.
* [x] In the following sequence, the pipette should arc for both movements. It should not move directly.

     1. Move to a well.
     2. `moveToCoordinates`.
     3. Move back to the original well.

* [ ] The handling of critical points is a little delicate and I only tenuously understand it.

    One case I know of where this matters is when an 8-Channel pipette is used, and the pipette starts off in a labware with the `centerMultichannelOnWells` quirk, like a reservoir or trough. Supposing the pipette got there with a `moveToWell` or similar, its initial "logical location" will be at the *center* of the well—even though the back-most channel is somewhere behind the center. If I got something wrong with critical points, the initial "move up" step of the arc will wrongly include an unexpected XY motion in addition to its normal Z motion.

# Other review requests

* [ ] Is the handling of critical points correct?
* [ ] Is the handling of critical points understandable? Can I explain it better through better comments, for example?
* [ ] As implemented, this will not dodge the Thermocycler like `moveToWell` does. This matches PAPIv2's behavior. We agree that in the future, we want more sophisticated collision checking based on geometry and coordinates and bounding boxes, rather than logical wells. **But for now:** Do we want to document anywhere that `moveToCoordinates` does not dodge the Thermocycler? Note that `moveToWell` does not currently document that it *does* dodge the Thermocycler.
* [ ] Is there any other location-tracking code I missed that needs to be adjusted?

# Risk assessment

Medium.

This isn't used by anything today as far as I know, so there's no risk to changing production behavior right now. But, it will be used in the future, and it is new motion planning code, which is sensitive. See the testing notes and review requests.

### CODE REVIEW

File: api/src/opentrons/motion_planning/waypoints.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ def get_waypoints(
     dest: Point,
     *,
     max_travel_z: float,
-    min_travel_z: float = 0.0,
+    min_travel_z: float,
```
Comment (SyntaxColoring) [inline api/src/opentrons/motion_planning/waypoints.py:21]: I made this argument mandatory because:

* Only one place was not providing an explicit value.
* I think we want to encourage callers to think about this value. It's likely unsafe to leave this at `0` because this needs to be high enough to clear any labware on the deck.

Comment (SyntaxColoring) [inline api/src/opentrons/motion_planning/waypoints.py:21]: I made this argument mandatory because:

* Only one place was not providing an explicit value.
* I think we want to encourage callers to think about this value. It's likely unsafe to leave this at `0` because this needs to be high enough to clear any labware on the deck.

File: api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py [hunk 2]
```diff
@@ -11,15 +15,84 @@
...
+
+    decoy.when(
+        hardware_api.get_instrument_max_height(mount=mount, critical_point=None)
+    ).then_return(max_height)
+
+    decoy.when(
+        state_view.motion.get_movement_waypoints_to_coords(
+            origin=current_position,
+            dest=Point(
+                params.coordinates.x, params.coordinates.y, params.coordinates.z
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py:77]: I keep going back and forth on whether it's a good idea for unit tests to reuse common values like this, versus copy-pasting a bunch of magic inline values like `Point(x=1.11, y=2.22, z=3.33)`.

On one hand, if every step redefines its own literals like `Point(x=1.11, y=2.21, z=3.13)`, things can get hard to read in longer tests like this one.

On the other hand, if we use common variables like this, the test is basically just a reimplementation of the production code.

File: api/tests/opentrons/protocol_engine/state/test_motion_view.py [hunk 3]
```diff
@@ -198,19 +217,18 @@ class WaypointSpec:
     all_labware_z: Optional[float] = None
     max_travel_z: float = 50
     should_dodge_thermocycler: bool = False
-    extra_waypoints: Sequence[Tuple[float, float]] = ()
+    extra_waypoints: Sequence[Tuple[float, float]] = field(default_factory=list)
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_engine/state/test_motion_view.py:208]: This change was necessary to work around a silly argument matching thing. The production code passes `[]` when there are no extra waypoints, not `()`. This discrepancy started to matter now that the tests are written to use `get_waypoints()` as a Decoy mock.

File: api/tests/opentrons/protocol_engine/state/test_motion_view.py [hunk 4]
```diff
@@ -198,19 +217,18 @@ class WaypointSpec:
...
 
 
-# TODO(mc, 2021-01-14): these tests probably need to be rethought; this fixture
-# is impossible to reason with. The test is really just trying to be a collaborator
-# test for the `get_waypoints` function, so we should rewrite it as such.
+# TODO(mm, 2022-06-22): Break this test apart into the parts that are orthogonal
+# (see this test's docstring) or refactor the subject to do less at once.
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocol_engine/state/test_motion_view.py:212]: Per discussion with @mcous, I've reworked this test a little bit.

Formerly, this test called the raw `get_waypoints()` function how it expected the *subject* to call it. Then, it compared the results from the raw call to the results from the subject.

Now, it sets up `get_waypoints()` as a monkey-patched mock. It can now assert that the subject calls `get_waypoints()` correctly through our usual Decoy mocking techniques.

I think this test is still hard to follow, but further refactors turned out to be a bit much for this PR, so I left it alone other than that change.

File: shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py [hunk 5]
```diff
@@ -43,7 +43,7 @@ class Params(BaseModel):
     flowRate: Optional[float]
     wellLocation: Optional[WellLocation]
     wait: Optional[int]
-    minimumZHeight: Optional[int]
+    minimumZHeight: Optional[float]
```
Comment (SyntaxColoring) [inline shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py]: To match the JSON protocol schema:

https://github.com/Opentrons/opentrons/blob/da09f5f490d51c79f4398b54164b542cfede78a8/shared-data/protocol/schemas/6.json#L420-L424

And our main Python model:

https://github.com/Opentrons/opentrons/blob/da09f5f490d51c79f4398b54164b542cfede78a8/api/src/opentrons/protocol_engine/commands/move_to_coordinates.py#L30

File: api/src/opentrons/protocol_engine/state/motion.py [hunk 6]
```diff
@@ -134,5 +128,50 @@ def get_movement_waypoints(
...
+            dest: The end point of the movement.
+            max_travel_z: How high, in deck coordinates, the pipette can go.
+                This should be measured at the bottom of whatever tip is currently
+                attached (if any).
+            direct: If True, move directly. If False, move in an arc.
+            additional_min_travel_z: The minimum height to clear, if moving in an arc.
+                Ignored if `direct` is True. If lower than the default height,
+                the default is used; this can only increase the height, not decrease it.
+        """
+        all_labware_highest_z = self._geometry.get_all_labware_highest_z()
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/motion.py:155]: This reliance on `get_all_labware_highest_z()` matches how `get_movement_waypoints_to_well()` works. But looking at this again, I think this might be a bug because it only takes labware into account, not modules.

Ticket this separately? Then we can:

* Double-check what PAPIv2 does
* Fix this (if necessary) for movements to coordinates and movements to wells
* Test it on a real robot with real modules

**Comment by SyntaxColoring** [comment]
Pretty sure I need to update this, too:

https://github.com/Opentrons/opentrons/blob/da09f5f490d51c79f4398b54164b542cfede78a8/api/src/opentrons/protocol_engine/state/pipettes.py#L74-L89

Edit: Yes, I definitely do. Without this, the pipette will incorrectly path though labware in the following sequence:

1. Move to well
2. Move to coordinates
3. Move to same well (💥)

Edit again: Done.

File: api/src/opentrons/protocol_engine/commands/move_to_coordinates.py [hunk 7]
```diff
@@ -1,13 +1,20 @@
...
 from typing_extensions import Literal
 
+from opentrons.types import Point
+
 from ..types import DeckPoint
 from .pipetting_common import PipetteIdMixin
 from .command import AbstractCommandImpl, BaseCommand, BaseCommandCreate
 
+if TYPE_CHECKING:
+    from opentrons.hardware_control import HardwareControlAPI
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/move_to_coordinates.py]: Does this need to be under the `TYPE_CHECKING` guard?

File: api/src/opentrons/protocol_engine/commands/move_to_coordinates.py [hunk 8]
```diff
@@ -55,12 +62,70 @@ class MoveToCoordinatesImplementation(
...
+    async def _move_to_coordinates(
+        self,
+        pipette_id: str,
+        deck_coordinates: DeckPoint,
+        direct: bool,
+        additional_min_travel_z: Optional[float],
+    ) -> None:
+        hw_mount = self._state_view.pipettes.get(
+            pipette_id=pipette_id
+        ).mount.to_hw_mount()
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/move_to_coordinates.py:96]: I'm down for a `get_hardware_mount` selector; I think I've seen this same think in a few places

File: api/src/opentrons/protocol_engine/state/motion.py [hunk 9]
```diff
@@ -1,15 +1,10 @@
 """Motion state store and getters."""
 from dataclasses import dataclass
-from typing import List, Optional
+from typing import List, Optional, Sequence, Tuple
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/motion.py]: `Sequence` and `Tuple` are unused which is causing linting failures

File: api/src/opentrons/protocol_engine/state/motion.py (hunk 6 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/state/motion.py:155]: Oh that's not good. PAPIv2 certainly takes all items into account (via the `Deck` class). Looks like we ignore modules if there's no labware on it. We'll want to resolve this, potentially in the 6.0.0 release. Will investigate today

File: api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py (hunk 2 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py:77]: When in doubt, I try to use value literals in test expectations

File: api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py [hunk 10]
```diff
@@ -11,15 +15,84 @@
...
+    decoy.verify(
+        *[
+            await hardware_api.move_to(
+                mount=mount,
+                abs_position=waypoint.position,
+                critical_point=waypoint.critical_point,
+            )
+            for waypoint in planned_waypoints
+        ]
+    )
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py]: Can we unroll this loop?

File: api/tests/opentrons/protocol_engine/state/test_motion_view.py (hunk 3 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/state/test_motion_view.py:208]: Oof. Hopefully `WaypointSpec` can go away in favor of the orthogonal tests you mention below

File: api/tests/opentrons/protocol_engine/state/test_motion_view.py [hunk 11]
```diff
@@ -25,24 +25,43 @@
...
+# The signature of the motion_planning.get_waypoints() function.
+_GET_WAYPOINTS_SIGNATURE = Callable[
+    ..., List[motion_planning.Waypoint]  # Parameter types not included, for simplicity.
+]
+
+
+@pytest.fixture
+def mock_get_waypoints(
+    decoy: Decoy, monkeypatch: pytest.MonkeyPatch
+) -> _GET_WAYPOINTS_SIGNATURE:
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_motion_view.py]: Given Python typing does not give us a convenient thing to say "give me the type signature of this function", I think we should:

- Change this to a `None` return and/or `autouse` fixture
- Use `decoy.when(motion_planning.get_waypoints(...))` in tests

Using `_GET_WAYPOINTS_SIGNATURE`, we lose all typechecking on the arguments of `get_waypoints` in tests, which increases the likelihood of miswritten stubs in tests (leading to falsely passing tests) and makes changing the signature of `get_waypoints` harder down the road because we won't have the typechecker guiding us to tests that need to change

File: api/tests/opentrons/protocol_engine/state/test_motion_view.py [hunk 12]
```diff
@@ -356,40 +425,161 @@ def test_get_movement_waypoints(
...
 ) -> None:
-    """It should raise FailedToPlanMoveError if get_waypoints raises."""
-    decoy.when(pipette_view.get_current_well()).then_return(None)
-    decoy.when(geometry_view.get_well_position("labware-id", "A1", None)).then_return(
-        Point(4, 5, 6)
+    """It should raise FailedToPlanMoveError if motion_planning.get_waypoints raises."""
+    decoy.when(
+        # TODO(mm, 2022-06-22): We should use decoy.matchers.Anything() for all
+        # arguments. For some reason, Decoy does not match the call unless we
+        # specify concrete values for all (?) arguments?
```
Comment (mcous) [inline api/tests/opentrons/protocol_engine/state/test_motion_view.py:437]: For the arguments that have default values that you don't care about in this test, you can try using `decoy.when(..., ignore_extra_args=True)`.

- [Usage guide](https://mike.cousins.io/decoy/usage/when/#only-specify-some-arguments)
- [API references](https://mike.cousins.io/decoy/api/#decoy.Decoy.when)

**Review by mcous — COMMENTED** [review]
Gonna try testing this out on a robot later today and add another review with testing results. Thanks for investigating `get_all_labware_highest_z`; it looks like a general ProtocolEngine bug to me.

Random thoughts, including some testing recommendations, below

**Review by TamarZanzouri — COMMENTED** [review]
Changes looks good to me! Mike gave some pretty good comments (that I would prob miss) and it looks like you fixed some of them already. once CI passes and you are done fixing @mcous comments would love to review this again

File: api/tests/opentrons/protocol_engine/state/test_motion_view.py (hunk 11 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/state/test_motion_view.py]: Good thinking.

**Review by mcous — COMMENTED** [review]
Tested on an OT-2R with a P20 Multi GEN2 and the endpoint behaved as expected (barring the fact that #10902 is not yet in `edge` / this branch.

Looks like we'll want to merge in `edge` to grab some CI fixes. From there, we can figure out what we want to do once @SyntaxColoring is back. Seems pretty much ready to go, barring some test tweaks

File: api/src/opentrons/protocol_engine/state/motion.py (hunk 6 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/state/motion.py:155]: Addressed in #10902.

**Comment by SyntaxColoring** [comment]
Re-drafting to fix this: https://github.com/Opentrons/opentrons/pull/10840#issuecomment-1164870264

File: api/src/opentrons/protocol_engine/commands/move_to_coordinates.py (hunk 7 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/commands/move_to_coordinates.py]: Nope! Removed.

File: api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py (hunk 2 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py:77]: How strongly do you feel about that, in this case?

[Here's how this test looks with literals](https://gist.github.com/SyntaxColoring/89078af6292e4179cebcfbd613ee553a). I personally find it pretty hard to follow the dataflow. Maybe this method has an unusually high number of values that have the same type? For example, there are at least 4 different `Point`s that ought to be thought of separately.

File: api/src/opentrons/motion_planning/waypoints.py [hunk 13]
```diff
@@ -41,17 +41,22 @@ def get_waypoints(
...
-    # `use_experimental_waypoint_planning` to True in
-    # `opentrons.protocols.geometry.plan_moves` causes three test failures at
-    # the time of this writing.
+    # NOTE(mm, 2022-06-22):
+    # This function is used by v6+ JSON protocols and v3+
+    # Python API protocols, but not v2 Python API protocols.
     #
-    # Eventually, it may take over for opentrons.hardware_control.util.plan_arc
+    # Flipping `use_experimental_waypoint_planning` to True to make PAPIv2 use this too
+    # causes three test failures at the time of this writing.
```
Comment (mcous) [inline api/src/opentrons/motion_planning/waypoints.py:49]: Just commenting to keep you up to date: I figured out at least two of the test failures! They're caused by behavioral differences in the waypoint planning:

- The old logic ignores `move_type: DIRECT` if you pass a `min_travel_z`, and arcs instead
    - Unclear to me if this behavior is actually used by calling code
    - But there is a test for it! So someone wanted it at some point
- The new logic suppresses duplicate waypoints
    - This behavior is probably unnecessary
    - In recent collaboration sessions, embedded suggested the idea of passing the waypoint list to the HW API rather than doing a loop in calling code. I like this idea, and it might make duplicate waypoint suppression even less important

File: api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py (hunk 2 above)
Comment (mcous) [reply api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py:77]: Not super strongly, just a rule of thumb to start from. If you find it easier to follow with the variables, then I think my only suggestion would be to name them `expected_whatever` whenever possible to make it clear that it's a variable we invented for the test

**Review by mcous — APPROVED** [review]
Good catch on wiping out the last known well
